crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -225,22 +235,165 @@ where
+        let (filter, reorg_blocks) = {
             let filters = self.inner.active_filters.inner.lock().await;
-            if let FilterKind::Log(ref filter) =
-                filters.get(&id).ok_or_else(|| EthFilterError::FilterNotFound(id.clone()))?.kind
-            {
-                *filter.clone()
+            let active_filter =
+                filters.get(&id).ok_or_else(|| EthFilterError::FilterNotFound(id.clone()))?;
+            if let FilterKind::Log(ref filter) = active_filter.kind {
+                (*filter.clone(), active_filter.reorg_blocks.clone())
             } else {
                 // Not a log filter
                 return Err(EthFilterError::FilterNotFound(id))
             }
         };
 
-        self.inner.logs_for_filter(filter).await
+        self.inner.logs_for_filter(filter, reorg_blocks).await
+    }
+}
+
+impl<Provider, Pool, Eth> EthFilter<Provider, Pool, Eth>
+where
+    Provider: BlockReader + BlockIdReader + EvmEnvProvider + 'static,
+    Pool: TransactionPool + 'static,
+    Eth: EthApiTypes + 'static,
+{
+    /// Create a background task and listener for reorged blocks updates relevant active filters.
+    /// In case of a chain reorg, previously emitted logs are emitted again but with the removed
+    /// field set to true.
+    pub fn spawn_watch_reorgs<Events>(&self, events: Events)
+    where
+        Events: CanonStateSubscriptions + 'static,
+    {
+        let this = self.clone();
+        self.inner.task_spawner.spawn_critical(
+            "eth-filters-watch-reorgs",
+            Box::pin(async move {
+                let notifications = events.canonical_state_stream();
+                this.watch_reorgs(notifications).await;
+            }),
+        );
+    }
+
+    /// Watch block reorgs and update filters accordingly.
+    async fn watch_reorgs(&self, mut notifications: CanonStateNotificationStream) {
+        while let Some(notification) = notifications.next().await {
+            if let CanonStateNotification::Reorg { old, .. } = notification {
+                self.update_reorg(old.blocks()).await;
+            }
+        }
+    }
+
+    /// Update a reorg block for all active filters.
+    async fn update_reorg(&self, old_blocks: &BTreeMap<BlockNumber, SealedBlockWithSenders>) {
+        let reorg_blocks: HashMap<BlockHash, SealedBlockWithSenders> =
+            old_blocks.iter().map(|(_, v)| (v.header.hash(), v.clone())).collect();
+
+        // we need to acquire a lock first
+        let mut active_filters = self.active_filters().inner.lock().await;
+
+        for active_filter in active_filters.values_mut() {
+            if let FilterKind::Log(filter) = &active_filter.kind {
+                match filter.block_option {
+                    FilterBlockOption::AtBlockHash(block_hash) => {
+                        if let Some(block) = reorg_blocks.get(&block_hash) {
+                            let block_num_hash = BlockNumHash::new(block.header.number, block_hash);
```

Comment:
could we document all of this a bit more clearly, because this is very hard to follow